x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java

```diff
@@ -33,6 +33,12 @@ public class SparseFileTracker {
      */
     private final TreeSet<Range> ranges = new TreeSet<>(RANGE_START_COMPARATOR);
 
+    /**
+     * Number of bytes from the start of the file that are present as one continuous range without gaps. Used as an optimization in
+     * {@link #waitForRange(ByteRange, ByteRange, ActionListener)} to bypass expensive inspection of {@link #ranges} in some cases.
+     */
+    private volatile long complete = 0L;
```

Comment:
> I wonder about the volatile here. It should work without it I think, since it only increases. Definitely requires a comment if we remove it...

Uff I guess maybe theoretically we could do that, but honestly, it's such a small performance difference and everyone ever looking at this code will wonder about whether this is actually safe. Seems to me we'd want to play it safe here if only for my nerves :D

Reply:
I am fine with that, just found it kind of interesting (and potentially safe, though I'd also want to spent a little more time on it if we went that route).